File: frontend/src/lib/passwordPolicy.ts
/** Must match backend/app/core/password_policy.py (ASCII-only rules). */

const RULES: ReadonlyArray<{
  readonly test: (password: string) => boolean;
  readonly message: string;
}> = [
  {
    test: (password) => password.length >= 8,
    message: "Password must be at least 8 characters long",
  },
  {
    test: (password) => password.length <= 128,
    message: "Password must be at most 128 characters long",
  },
  {
    test: (password) => /[A-Z]/.test(password),
    message: "Password must contain at least one uppercase letter",
  },
  {
    test: (password) => /[a-z]/.test(password),
    message: "Password must contain at least one lowercase letter",
  },
  {
    test: (password) => /\d/.test(password),
    message: "Password must contain at least one digit",
  },
];

export function getPasswordErrors(password: string): string[] {
  return RULES.filter((rule) => !rule.test(password)).map(
    (rule) => rule.message,
  );
}

export function isPasswordValid(password: string): boolean {
  return getPasswordErrors(password).length === 0;
}

File: backend/app/core/password_policy.py
"""Shared password strength rules for register and reset."""

from __future__ import annotations


def get_password_errors(password: str) -> list[str]:
    """Return human-readable validation errors (empty if password is valid)."""
    errors: list[str] = []
    if len(password) < 8:
        errors.append("Password must be at least 8 characters long")
    if len(password) > 128:
        errors.append("Password must be at most 128 characters long")
    if not any(c.isupper() for c in password):
        errors.append("Password must contain at least one uppercase letter")
    if not any(c.islower() for c in password):
        errors.append("Password must contain at least one lowercase letter")
    if not any(c.isdigit() for c in password):
        errors.append("Password must contain at least one digit")
    return errors


def validate_password_strength(password: str) -> str:
    """Return the password if valid; raise ValueError with the first rule violation."""
    errors = get_password_errors(password)
    if errors:
        raise ValueError(errors[0])
    return password
